archiver/worker.go

```diff
@@ -0,0 +1,264 @@
+start:
+
+	jobs, limitReached, err := w.getJobs()
+	if err != nil {
+		log.Errorw("failed to fetch jobs for archiving", "error", err)
+		return false
+	}
+
+	if len(jobs) == 0 {
+		return false
+	}
+
+	if !limitReached && time.Since(w.lastUploadTime) < w.config.uploadFrequency {
+		return false // respect the upload frequency
+	}
+
+	workspaceID := jobs[0].WorkspaceId
+	log = log.With("workspaceID", workspaceID)
+	storagePrefs, err := w.storageProvider.GetStoragePreferences(workspaceID)
+	if err != nil {
+		log.Errorw("failed to fetch storage preferences", "error", err)
+		if err := w.markStatus(
+			jobs,
+			jobsdb.Aborted.State,
+			errJson(err),
+		); err != nil {
+			if w.lifecycle.ctx.Err() != nil {
+				return false
+			}
+			log.Errorw("failed to mark unconfigured archive jobs' status", "error", err)
+			panic(err)
+
+		}
+		if !limitReached {
+			return true
+		}
+		goto start
+	}
+	if !storagePrefs.Backup(w.archiveFrom) {
+		if err := w.markStatus(
+			jobs,
+			jobsdb.Aborted.State,
+			errJson(fmt.Errorf("%s archival disabled for workspace %s", w.archiveFrom, workspaceID)),
+		); err != nil {
+			if w.lifecycle.ctx.Err() != nil {
+				return false
+			}
+			log.Errorw("failed to mark archive disabled jobs' status", "error", err)
+			panic(err)
+		}
+		if !limitReached {
+			return true
+		}
+		goto start
+	}
+
+	location, err := w.uploadJobs(w.lifecycle.ctx, jobs)
+	if err != nil {
+		log.Errorw("failed to upload jobs", "error", err)
+		return false
+	}
+	w.lastUploadTime = time.Now()
+
+	if err := w.markStatus(
+		jobs,
+		jobsdb.Succeeded.State,
+		locationJson(location),
+	); err != nil {
+		if w.lifecycle.ctx.Err() != nil {
+			return false
+		}
+		log.Errorw("failed to mark successful upload status", "error", err)
+		panic(err)
+	}
+	if !limitReached {
+		return true
+	}
+	goto start
+}
+
+func (w *worker) SleepDurations() (min, max time.Duration) {
+	if w.lastUploadTime.IsZero() {
+		return w.config.minSleep, w.config.uploadFrequency
+	}
+	return w.config.minSleep, time.Until(w.lastUploadTime.Add(w.config.uploadFrequency))
+}
+
+func (w *worker) Stop() {
+	w.lifecycle.cancel()
+}
+
+func (w *worker) uploadJobs(ctx context.Context, jobs []*jobsdb.JobT) (string, error) {
+	defer w.uploadLimiter.Begin(w.sourceID)()
+	firstJobCreatedAt := jobs[0].CreatedAt.UTC()
+	lastJobCreatedAt := jobs[len(jobs)-1].CreatedAt.UTC()
+	workspaceID := jobs[0].WorkspaceId
+	log := w.log.With("workspaceID", workspaceID)
+
+	gzWriter := fileuploader.NewGzMultiFileWriter()
+	path := fmt.Sprintf(
+		"%v%v.json.gz",
+		lo.Must(misc.CreateTMPDIR())+"/rudder-backups/"+w.sourceID+"/",
+		fmt.Sprintf("%v_%v_%v", firstJobCreatedAt.Unix(), lastJobCreatedAt.Unix(), workspaceID),
+	)
+
+	for _, job := range jobs {
+		j, err := marshalJob(job)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal job: %w", err)
+		}
+		if _, err := gzWriter.Write(path, append(j, '\n')); err != nil {
+			return "", fmt.Errorf("failed to write to file: %w", err)
```

Comment:
[minor] no need to mention `failed to` on an error message
```suggestion
			return "", fmt.Errorf("write to file: %w", err)
```